### PR TITLE
Firefox: pass PageAuthenticateTests

### PR DESCRIPTION
**Pull request by Meir017**
related to #106

### CODE REVIEW

File: src/PlaywrightSharp.Firefox/FirefoxPage.cs [hunk 1]
```diff
@@ -246,7 +250,13 @@ public Task SetCacheEnabledAsync(bool enabled)
 
         public Task SetRequestInterceptionAsync(bool enabled) => throw new NotImplementedException();
 
-        public Task AuthenticateAsync(Credentials credentials) => throw new NotImplementedException();
+        public Task AuthenticateAsync(Credentials credentials)
+            => _session.SendAsync(
+                new NetworkSetAuthCredentialsRequest
+                {
+                    Username = credentials?.Username,
```
Comment (Meir017) [inline src/PlaywrightSharp.Firefox/FirefoxPage.cs:257]: we need to allow to serialize as `{ username: null, password: null }`

Comment (kblok) [reply src/PlaywrightSharp.Firefox/FirefoxPage.cs:257]: Can you explain this a little bit more?

File: src/PlaywrightSharp.Firefox/FirefoxPage.cs [hunk 2]
```diff
@@ -494,7 +504,7 @@ private void OnWorkerCreated(PageWorkerCreatedFirefoxEvent e)
             string workerId = e.WorkerId;
             var worker = new Worker(e.Url);
             FirefoxSession tempWorkerSession = null;
-            var workerSession = new FirefoxSession(_session.Connection, "worker", workerId, async (id, request) =>
+            var workerSession = new FirefoxSession(_session.Connection, "worker", workerId, async (id, request, jsonOptions) =>
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxPage.cs:507]: Same, could you explain this?

File: src/PlaywrightSharp.Firefox/FirefoxPage.cs (hunk 1 above)
Comment (Meir017) [reply src/PlaywrightSharp.Firefox/FirefoxPage.cs:257]: we need to send a json that includes both properties with a null value, currently we ignore all null values

File: src/PlaywrightSharp.Firefox/FirefoxPage.cs (hunk 2 above)
Comment (Meir017) [reply src/PlaywrightSharp.Firefox/FirefoxPage.cs:507]: this is in order for us to be able to use custom options that will allow us to handle properties in a special way

File: src/PlaywrightSharp.Firefox/FirefoxPage.cs [hunk 3]
```diff
@@ -246,7 +250,13 @@ public Task SetCacheEnabledAsync(bool enabled)
...
         public Task SetRequestInterceptionAsync(bool enabled) => throw new NotImplementedException();
 
-        public Task AuthenticateAsync(Credentials credentials) => throw new NotImplementedException();
+        public Task AuthenticateAsync(Credentials credentials)
+            => _session.SendAsync(
+                new NetworkSetAuthCredentialsRequest
+                {
+                    Username = credentials?.Username,
+                    Password = credentials?.Password,
+                }, options: _authenticateJsonOptions);
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxPage.cs:259]: We don't need the `_authenticateJsonOptions` variable, we could create the instance here. It won't be that expensive and we won't have that in memory all the time.

File: src/PlaywrightSharp.Firefox/FirefoxPage.cs (hunk 1 above)
Comment (kblok) [reply src/PlaywrightSharp.Firefox/FirefoxPage.cs:257]: Can't we mark Username and Password to send null all the time? 
I know it's autogenerated, but can we think about a workaround there? I would rather mess with the code generator than adding a boilerplate only for this.

File: src/PlaywrightSharp.Firefox/FirefoxPage.cs [hunk 4]
```diff
@@ -1,617 +1,616 @@
...
-
-        private class AuxData
-        {
-            public string FrameId { get; set; }
-
-            public string Name { get; set; }
-        }
-    }
-}
+using System;
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxPage.cs:1]: Ouch

File: src/PlaywrightSharp.Firefox/FirefoxPage.cs (hunk 1 above)
Comment (Meir017) [reply src/PlaywrightSharp.Firefox/FirefoxPage.cs:252]: hmm, I'm not sure how. I could try adding a `JsonConverter` for this class

Comment (Meir017) [reply src/PlaywrightSharp.Firefox/FirefoxPage.cs:252]: once #477 is merged I'll add the JsonConverter for this

Comment (kblok) [reply src/PlaywrightSharp.Firefox/FirefoxPage.cs:252]: #477 Merged.

File: src/PlaywrightSharp.Firefox/Helper/NetworkSetAuthCredentialsRequestJsonConverter.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+using PlaywrightSharp.Firefox.Protocol.Network;
+using PlaywrightSharp.Firefox.Protocol.Runtime;
+using PlaywrightSharp.Helpers;
+
+namespace PlaywrightSharp.Firefox.Helper
+{
+    /// <summary>
+    /// We need to be able to send null values in order to reset the username/password values.
+    /// </summary>
+    internal class NetworkSetAuthCredentialsRequestJsonConverter : JsonConverter<NetworkSetAuthCredentialsRequest>
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/Helper/NetworkSetAuthCredentialsRequestJsonConverter.cs:13]: My concern here is whether we are going to need more of these or not. I wouldn't like to have 5 classes like this.

Comment (Meir017) [reply src/PlaywrightSharp.Firefox/Helper/NetworkSetAuthCredentialsRequestJsonConverter.cs:13]: I think this is the last one

**Pull request merged**